src/Emulator/Peripherals/Peripherals/MTD/OpenTitan_FlashController.cs

```diff
@@ -0,0 +1,1185 @@
+                infoFlash.WriteByte(pageOffset + i, 0xff);
+            }
+        }
+
+        private void DoReadData(long busWordAddress){
+            this.Log(
+                LogLevel.Noisy,
+                "OpenTitan_FlashController/DoReadData: busWordAddress = 0x{0:X}",
+                busWordAddress);
+            uint value = flash.ReadDoubleWord(busWordAddress - FLASH_MEM_BASE_ADDR);
+            this.Log(
+                LogLevel.Noisy,
+                "OpenTitan_FlashController/DoReadData: value = 0x{0:X}",
+                value);
+            if(IsOperationAllowed(OperationType.READ_DATA, busWordAddress)){
+                readFifo.Enqueue(value);
+            }else{
+                interruptStatusRegisterOpErrorFlag.Value |= true;
+                opStatusRegisterErrorFlag.Value |= true;
+            }
+            this.Log(
+                LogLevel.Noisy,
+                "OpenTitan_FlashController/DoReadData: readFifo.Count = {0}",
+                readFifo.Count);
+        }
+
+        private void DoReadInfo(long busWordAddress){
+            /*
+            ** TODO(julianmb): memory protection should be implemented.
+            ** It currently doesnt have a unit test though.
+            ** BANK*_INFO*_PAGE_CFG_* registers are of interest
+            */
+            this.Log(
+                LogLevel.Noisy,
+                "OpenTitan_FlashController/DoReadInfo: busWordAddress = 0x{0:X}",
+                busWordAddress);
+            uint value = infoFlash.ReadDoubleWord(busWordAddress - FLASH_MEM_BASE_ADDR);
+            this.Log(
+                LogLevel.Noisy,
+                "OpenTitan_FlashController/DoReadInfo: value = 0x{0:X}",
+                value);
+            readFifo.Enqueue(value);
+            this.Log(
+                LogLevel.Noisy,
+                "OpenTitan_FlashController/DoReadInfo: readFifo.Count = {0}",
+                readFifo.Count);
+        }
+
+        private void DoProgramData(long busWordAddress){
+            this.Log(
+                LogLevel.Noisy,
+                "OpenTitan_FlashController/DoProgramData: busWordAddress = 0x{0:X}",
+                busWordAddress);
+            uint value = programFifo.Dequeue();
+            this.Log(
+                LogLevel.Noisy,
+                "OpenTitan_FlashController/DoProgramData: value = 0x{0:X}",
+                value);
+            if(IsOperationAllowed(OperationType.PROGRAM_DATA, busWordAddress)){
+                flash.WriteDoubleWord(busWordAddress - FLASH_MEM_BASE_ADDR, value);
+            }else{
+                interruptStatusRegisterOpErrorFlag.Value |= true;
+                opStatusRegisterErrorFlag.Value |= true;
+            }
+
+            this.Log(
+                LogLevel.Noisy,
+                "OpenTitan_FlashController/DoProgramData: programFifo.Count = {0}",
+                programFifo.Count);
+        }
+
+        private void DoProgramInfo(long busWordAddress){
```

Comment:
I suggest extracting the common logic to a helper method (vide `DoProgramData`).